### ISSUE TITLE
Integrating attention and feature vectors?

### ISSUE THREAD

**Issue by Centaurea-Platinum** [lightvector/KataGo#1080]
I want to train my own net and make some additions onto the katago framework.
Using my very limited AI knowledge, I theorized that a 2d pos encoded attention layer could perhaps reduce the need for deep nets and training costs on learning global relationships? 
Furthermore, could returning and updating feature vectors in the MCTS allow for better propagation of data in the tree and thus better game play? 

For instance in the following framework:
Evaluation
		Input[m,n,3]: ToPlay(full plane of 0,1) + illegal move mask(0,1) + current board state(-1, 0, 1) 
		Input[r]: Specific rules(binary encoded: AreaTerritory, SekiScoring, TaxRule,...) and komi
		Output[m,n,3]: probability that a certain position is the best move, force positions on illegal move mask to be 0 + expected winrate + probability ownership of position is black
		NeuralNet
			Pad input
			[m+1,n+1,3] input through resnet, transform into [m+1,n+1,1024] stack of feature map
			[m+1,n+1,1024] stack of feature maps through stacked layers of multi head attention
			[m+1,n+1,1024]  through multi head attention pool into [1024]
			[r] set of rules and komi into [1024] vector through MLP
			use the two [1024] vector as weight for the feature map, multiplying each feature map with its two corresponding weights(late fusion of specifications)
			resnet, decrease dimensions of feature map from [m+1,n+1,1024] into [m+1,n+1,256]
			use multi head attention pooling to combine the [m+1,n+1,256] tensor into a [256] vector to return for the MCTS tree
			resnet, decrease [m+1,n+1,256] further into the output of [m,n,3], spatial dimensions is transformed in the last convolution layer by not padding (256->64->3)

		MCTS
			Assign weights based on best move prob + noise, implement force playout
			returns the [256] vector to update the [m+1,n+1,256] tensor at the corresponding position

**Comment by lightvector** [comment]
If you can find something better it would be very cool.

I've tried vanilla self-attention once or twice before but the challenge is that it's expensive for a 19x19 board, and when I tried it I recall it made the net enough slower that it if you measured prediction accuracy/loss holding inference cost constant (i.e. training a couple sizes and comparing the accuracy/speed tradeoff curve to the curve for convolutional resnets), it was worse. There might be ways to optimize it, adjust training hyperparameters, etc. and make it so that it's faster or better. That's the entire challenge though. There are endless published papers with different architectures, and different "efficient attention" mechanisms, and so on to try, so there's no shortage of ideas, the entire bottleneck and value would come from empirically testing such ideas and figuring out if there is any way to make them fast enough to be competitive, if you're interested to experiment with it.

If you want to experiment with passing feature data through MCTS you'll run need to figure out how to train it, the system becomes very RNN-like. Normally for an RNN-like system to be effective you want to backprop through the entire time-sequence, and doing that increases the difficulty of coming up with a stable way to train it and implementing how the data is produced and written down and so on, and also the way the system works becomes more entangled with the implementation of the search, rather than them being modular and tunable and testable independently. There's a reason why there are a huge number of "AlphaZero-like" projects and replications out there, but relatively fewer "MuZero-like" things out there (even without the learned world model part, even only implement the part where it's recurrent through the search). It's harder to implement and test and train. If want to figure out make it work though and were able to show that it could be better at equal compute and that continues to hold as you scale it up (e.g. millions of playouts), that would also be a worthy accomplishment.

**Comment by Allium-altaicum** [comment]
@Centaurea-Platinum There are some different layers to mimic the large receptive field of attention, for example, [WTConv](https://arxiv.org/abs/2407.05848), its computational cost isn't high.
